### ISSUE TITLE
IndexedDB

### ISSUE THREAD

**Issue by sh7dm** [denoland/deno#1699]
It'll be useful to have IndexedDB in Deno.
Related to #1657.

**Comment by fengmu456** [comment]
Should we use pure TS implementation or on the Rust side?

**Comment by fengmu456** [comment]
Or use ready-made C++ implementations
https://cs.chromium.org/chromium/src/content/browser/indexed_db/

**Comment by kitsonk** [comment]
This is really complicated when you dig into it.

There is no Rust implementation of IndexDB (there is a crate that appears to be squatting, but it really isn't an implementation).  Servo does not support IndexedDB, they have a long standing issue to implement.  So on the Rust front it is sort of a dead end.

In Node.js, there are no realistic implementations either.  Again there is a package squatting on indexeddb, but it isn't really a good package at all.  The one thing that seems to have promise is [IndexedDBShim](https://www.npmjs.com/package/indexeddbshim) which sits on top of WebSQL, and there are native implementations of WebSQL on Node.js using sqlite3.

jsdom doesn't implement IndexedDB or WebSQL, though there is [this issue](https://github.com/jsdom/jsdom/issues/1748) talking about implementing the IndexedDB shim mentioned above, I assume requiring the WebSQL implementation.  It looks like it is stalled.

There is also an in memory [IndexedDB API written in TypeScript](https://github.com/dumbmatter/fakeIndexedDB).

There is what looks like a realistic Rust create for [sqlite3](https://docs.rs/rust-sqlite/0.3.0/sqlite3/).

So outside of using a C++ implementation, which would be a break in the pattern, the only realistic path I can easily see is that to use the in memory IndexedDB written in TypeScript, which would not give persistent storage, or we could adapt the in memory version and wire it up to a Rust sqlite3 bindings to persist the data.

I don't think just persisting out to JSON really works, because IndexedDB is a bit more complex than that, it contains concepts like transactions and other database features which is why almost all the browser implementations are based on sqlite3 IIRC.

**Comment by kitsonk** [comment]
Just updating this, servo still doesn't have it, and Gecko is looking at refactoring, but it is sqlite3 under the hoods.  Here is a the most recent post that lays things out most relevantly: https://github.com/servo/servo/issues/6963#issuecomment-528852316.

It would be really good to get localStorage/IndexedDB/kv-store in to Deno, IMO.

**Comment by anlexN** [comment]
deno embrace web standard, indexeddb is one of web standard. so deno **must have** indexeddb!
@ry ,Are you going to ignore it again? like you ignore many things about nodejs.

**Comment by kitsonk** [comment]
@anlexN your comment is borderline abusive.  The issue is well discussed here of how it is a complex issue.  Please keep your conversation civil and on-topic.

**Comment by jacobbogers** [comment]
HI, just exploring deno a bit, i thought (my mental model) is that deno implemented chromium aswell, (so it 100% compatible (as much as possible) with all webapis)
The index db source code is here, it is in C though
https://github.com/chromium/chromium/tree/99314be8152e688bafbbf9a615536bdbb289ea87/third_party/google-closure-library/closure/goog/db

**Comment by kitsonk** [comment]
@jacobbogers Deno does not implement Chromium, it only implements V8. All the Web APIs need to be implemented independently. Chromium tends to implement these in C++ using the v8 binding to inject objects into the isolate.

Yes, both Firefox and Chromium are open source and their code is available, but that doesn't make it magically work for Deno.

**Comment by jacobbogers** [comment]
> @jacobbogers Deno does not implement Chromium, it only implements V8. All the Web APIs need to be implemented independently. Chromium tends to implement these in C++ using the v8 binding to inject objects into the isolate.
> 
> Yes, both Firefox and Chromium are open source and their code is available, but that doesn't make it magically work for Deno.

thanks for explaining, well at least if the source is avail (Firefox/Chrome) then we can kindof get some ideas /inspiration how to implement it,

 IndexDB is a tuff nut to crack, it has transactions but has no explicit commit (its weird and ingenious way of doing transactions) this would not be easy to build from scratch.

**Comment by kitsonk** [comment]
@jacobbogers we have ideas/inspiration.  One of the biggest factors was that we didn't have easy access to the V8 serialize/deserialize routines, which you need to support "structured cloning" which is needed to be able to store objects.  We have that now.  The other problem is the underlying DB as well as support under Rust.  LevelDB (the one Google created and is in Chromium) is a bit of a mess. Firefox and Safari use SQLite as well as Servo looked at SQLite (can't remember how far they got).  The biggest problem is the analogues between how Chromium and Firefox operate, and how Deno operates mean that there is very little overlap.  For implementing web standards, we have had a lot better luck looking at a JavaScript shim/polyfill and the specification itself, than we have looking at other browser implementations.

There are the wpt for IndexedDB which helps ensure the standard is implemented correctly, and we now have the wpt integrated into our test process.

**Comment by brainless** [comment]
Hey everyone just reading the discussions. I think IndexDB for Deno *might* end up being used in more varied (or larger) ways than in a browser. For example, it is OK to expect data eviction in the browser, but if developers use IndexDB as a persistent backend store (like say SQLite) then the they might want further control of the underlying storage, its space allocation, path, etc.. Ability to backup/restore also comes to mind.

Maybe my thought process is incorrect, just I felt it is perhaps a good time to share them now, before we have a ready implementation.

**Comment by kitsonk** [comment]
> then the they might want further control of the underlying storage, its space allocation, path, etc..

Then that isn't IndexedDB, that is a native SQLite plugin.

**Comment by brainless** [comment]
> > then the they might want further control of the underlying storage, its space allocation, path, etc..
> 
> Then that isn't IndexedDB, that is a native SQLite plugin.

You are right @kitsonk this is not IndexDB. I am just saying that people *might* end up using IndexDB in a different way. Maybe just great documentation will help in case people start using IndexDB as a traditional backend store.

**Comment by tracker1** [comment]
Small input on this one, investigating leveldb as a potential underlying option to implement IndexedDB on top of, and it looks like most of the rust crates are just bindings to the C library for leveldb and the single rust native implementation hasn't been updated in a long time with a warning.

I do think a rust native leveldb implementation with an IndexedDB compatibility layer over it would be a very good approach.  Though, there aren't enough roots planted in that space (yet).  Not sure if there is any interest in implementing such a thing in native rust, exposing a C api that matches the existing library for testing, well outside the scope of Deno.

**Comment by crowlKats** [comment]
I am currently implementing IndexedDB. It will be based on SQLite as that was the decision that was made when picking the backend for WebStorage.

**Comment by Platzer** [comment]
> > > then the they might want further control of the underlying storage, its space allocation, path, etc..
> > 
> > 
> > Then that isn't IndexedDB, that is a native SQLite plugin.
> 
> You are right @kitsonk this is not IndexDB. I am just saying that people _might_ end up using IndexDB in a different way. Maybe just great documentation will help in case people start using IndexDB as a traditional backend store.


Think about application Migration in case of a multi application server, it would be nice to just copy a single file instead of extracting a table from a file in some sys directory. 🧐

**Comment by crowlKats** [comment]
@Platzer well the idb file will be accessible just like the localstorage file.

**Comment by Platzer** [comment]
> @Platzer well the idb file will be accessible just like the localstorage file.

Sorry I hadn't time to try it out, found it with `deno info`. This should work fine, thx 💪

**Comment by petamoriken** [comment]
Unfortunately, IndexedDB is not useful at the moment. Specifically, transactions cannot be fully wrapped in `Promise` because they are auto-committed after the all requests and microtasks are completed. This should be solved by a new API in the future, but we don't know when that will be.

https://github.com/w3c/IndexedDB/issues/34#issuecomment-533354259

**Comment by johnspurlock** [comment]
> IndexedDB is not useful at the moment

Just wanted to say that while IndexedDB is not ideal, it is most certainly useful, and would be very useful in Deno for writing isomorphic browser/server-side libraries and apps.

**Comment by janus-reith** [comment]
> I am currently implementing IndexedDB. It will be based on SQLite as that was the decision that was made when picking the backend for WebStorage.

Hey @crowlKats, I think support for IndexedDB in Deno would be awesome, have you been able to make progress on this, or did you face any major roadblocks?

Btw hope I won't get too far off topic with this, but with new architecture options like [Cloudflare Workers](https://developers.cloudflare.com/workers/learning/using-durable-objects) emerging, and hoping Deno Deploy would eventually add such a feature too, there could be a lot of use cases for  IndexedDB on the server side. (Although I'm not sure wether storage options with a a standard-compliant implementation would actually be flexible enough for use cases I have in mind.)

**Comment by crowlKats** [comment]
> Hey @crowlKats, I think support for IndexedDB in Deno would be awesome, have you been able to make progress on this, or did you face any major roadblocks?

I havent. A bit after i wrote that comment, i stopped, as:
A) i didnt have the time for such a complex api
B) the api sucks to either make or use
C) I am not all too familiar with database related things
Maybe at some point i will pick it up again

Regarding storage on Deploy:
Having a KV storage is something that we are considering

**Comment by janus-reith** [comment]
> I havent. A bit after i wrote that comment, i stopped, as: A) i didnt have the time for such a complex api B) the api sucks to either make or use C) I am not all too familiar with database related things Maybe at some point i will pick it up again
> 
> Regarding storage on Deploy: Having a KV storage is something that we are considering

Thanks for the quick response!

**Comment by aaronhuggins** [comment]
Anyone looking for a polyfill, I've got [IndexedDBShim working here](https://github.com/aaronhuggins/pouchdb_deno/blob/main/modules/indexeddb/mod.ts) based on a [WebSQL ponyfill here](https://github.com/aaronhuggins/pouchdb_deno/blob/main/modules/websql/mod.ts).

**Comment by jsejcksn** [comment]
@crowlKats I think having KV storage is a more accessible first goal than an IDB implementation. Have you considered this idea, and, if so, what attributes would you expect such a feature to have? Here are some examples of what I mean by that:

- each store is namespaced (many kv stores, each accessed by its ID/name)
- each store's API is [map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)-like
- (even if structured clone or circular references would be an issue) at least all JSON types should be supported as values, i.e.
    ```ts
    type Json = boolean | number | null | string | Json[] | { [key: string]: Json };
    type Value = Json | undefined;
    ```
- async?
- backed by SQLite?

**Comment by crowlKats** [comment]
But that has to do with web compat, or rather, in this case, incompatibility. What we put in the deno ns and what we dont is carefully considered based on if there is a web api that could take care of it.

**Comment by jsejcksn** [comment]
I think I now understand the miscommunication (and perhaps I didn't make it clear in my initial message): my suggestion is entirely unrelated to web compatibility. It's simply about offering a built-in API for persistent KV storage of native data types.

**Comment by crowlKats** [comment]
Yes, I understand that, but we wont go that route, since there is a web compatible solution: IDB. having 2 different APIs on a later date would be suboptimal; I wouldnt consider this as an option. KV-Storage is a dead spec. We have had this kind of discussion multiple times internally already, and we havent reached a final conclusion yet

**Comment by jsejcksn** [comment]
> We have had this kind of discussion multiple times internally already, and we havent reached a final conclusion yet

Are there published notes from those? (I was actually searching for kv storage and that's how I found this issue.)

**Comment by crowlKats** [comment]
No, since this is an ongoing discussion

**Comment by aaronhuggins** [comment]
@crowlKats I have some interest in implementing this, but I don't want to take something away that you really want to do. If it *is* okay, should I start from your branch in #14035? Or just cut a fresh branch independent of yours and open a separate PR?

**Comment by aaronhuggins** [comment]
@crowlKats I have some interest in implementing this, but I don't want to take something away that you really want to do. If it *is* okay, should I start from your branch in #14035? Or just cut a fresh branch independent of yours and open a separate PR?

**Comment by crowlKats** [comment]
@aaronhuggins The PR isnt too far from completion, but still a bit of work left. thing is, it currently isnt high priority, so leaving this on the backburner for a bit. I would advise against starting your own PR

**Comment by jsejcksn** [comment]
Ref: https://github.com/denoland/deno/pull/18232

Is this groundwork for a persistent storage API in Deno Deploy?

**Comment by bartlomieju** [comment]
@jsejcksn that's correct.

**Comment by idanran** [comment]
> @crowlKats I have some interest in implementing this, but I don't want to take something away that you really want to do. If it _is_ okay, should I start from your branch in #14035? Or just cut a fresh branch independent of yours and open a separate PR?

I'm kind of looking forward to it.

**Comment by alexgleason** [comment]
Now that Deno.KV is a thing, would it make sense to make IndexedDB a wrapper around the Deno.KV API instead of yet another SQLite? That way IndexedDB on Deno Deploy would take advantage of FoundationDB, and it would mean improvements to Deno.KV locally (see: https://github.com/denoland/deno/issues/18657) would improve the IndexedDB implementation.

**Comment by idofilus** [comment]
I needed this support for creating unit testing designed for the browser and tested by deno.
So this worked for me:

```typescript
import { assertEquals } from "@std/assert";
import { indexedDB } from "npm:fake-indexeddb@6.0.0";

globalThis.indexedDB = indexedDB;

// The actually implementation of the indexedDB APIs 
import LocalDataSource from "../../app/src/core/LocalDataSource.ts";

Deno.test(async function LocalDataSourceTest()
{
    assertEquals(1, 1);
});
```

**Comment by mfulton26** [comment]
@idofilus, as far as I know, `npm:fake-indexeddb` does not persist to disk which is something that would be desired from Deno built-in support. The more Deno runtime works like browser runtimes the easier it is to write universal code and the less APIs there are to learn and code against.

**Comment by bartlomieju** [comment]
It's 2026 and most people use some sort of SQLite flavor. We have native support for 
 `node:sqlite` and also `Deno.KV` which is built on top of SQLite. We're not gonna do IndexedDB after all.

**Comment by alexgleason** [comment]
This is one of the few things in 2026 that could make Deno more appealing than Bun

**Comment by tracker1** [comment]
@bartlomieju  I think having IndexedDB is important and should remain as a feature request for browser parity.

I don't know what it would take to implement either on top of SQLite or via LevelDB, but as a JS/TS engine feature, I think it is a pretty wide miss at this point.

**Comment by bartlomieju** [comment]
I'm gonna reopen this for now, but there's a slim chance we will implement this.

What do people want to use it for anyway?

**Comment by idofilus** [comment]
> I'm gonna reopen this for now, but there's a slim chance we will implement this.
> 
> What do people want to use it for anyway?

Another major browser api that can be used in the code designed to run on browsers, with access to the all other of deno apis.
In my [given case](https://github.com/denoland/deno/issues/1699#issuecomment-2764815948) provided in this thread, I have a `DataSource` core class handling database related connection, e.g. mongodb `MongoDataSource`. Which I also allow the same core class as local storage, e.g. in the browser - this for example will be an offline database, with the option to sync against the remote e.g. mongo.

Since my server side is in Deno, and having that server side only code e.g. `MongoDataSource` I have the unit testing in it, now since it's the same `DataSource` core class, I made the unit testing on that class, and just creating the instances for the other inherited classes, therefore running a unit testing in deno to validate the browser's side `LocalDataSource` in the same pipeline.

The more browser apis in deno, the better "shared code" we can have, IndexedDB is quite useful for many use cases where compatibility of the code with the browser is required.

**Comment by alexgleason** [comment]
> What do people want to use it for anyway?

I want to test my browser libraries.

**Comment by oscarotero** [comment]
Having support for IndexDB allows to write code that works on both front and backend and use libraries like Dexie.js or localForage that work on top of indexDB. It does a similar job as Deno.KV, but it's more standard, with more features and a bigger ecosystem. In fact, I think Deno.KV could be a `@std` library built on top of indexDB.
